core/src/components/modal/modal.tsx

```diff
@@ -345,10 +345,47 @@ export class Modal implements ComponentInterface, OverlayInterface {
   }
 
   componentWillLoad() {
-    const { breakpoints, initialBreakpoint, el } = this;
+    const { breakpoints, initialBreakpoint, el, htmlAttributes } = this;
     const isSheetModal = (this.isSheetModal = breakpoints !== undefined && initialBreakpoint !== undefined);
 
-    this.inheritedAttributes = inheritAttributes(el, ['aria-label', 'role']);
+    const attributesToInherit = ['aria-label', 'role'];
+    this.inheritedAttributes = inheritAttributes(el, attributesToInherit);
+
+    /**
+     * When using a controller modal you can set attributes
+     * using the htmlAttributes property. Since the above attributes
+     * need to be inherited inside of the modal, we need to look
+     * and see if these attributes are being set via htmlAttributes.
+     *
+     * We could alternatively move this to componentDidLoad to simplify the work
+     * here, but we'd then need to make inheritedAttributes a State variable,
+     * thus causing another render to always happen after the first render.
```

Comment:
I'm okay with this approach. There's probably a cleaner way of doing it with some clever use of `Object.keys()`/`filter()`/`reduce()`, but the general strategy works for me. I came up with this after staring at it for 20 minutes; might be slightly easier to grok and it cuts a layer of nesting.
```ts
if (htmlAttributes !== undefined) {
  const htmlAttributesToInherit = Object.keys(htmlAttributes).filter(attr => attr in attributesToInherit);
  htmlAttributesToInherit.forEach((attribute) => {
    this.inheritedAttributes = {
      ...this.inheritedAttributes,
      [attribute]: htmlAttributes[attribute],
    };

    delete htmlAttributes[attribute];
  });
}

Reply:
I ended up just sticking with the approach in the PR. The `Object.keys` approach has some scalability concerns since worst case you'll need to iterate through `htmlAttributes` 3 times. (Object.keys, then filtering the array, then iterating through the result). We only inherit 2 attributes, so the performance impact right now is negligible. But if we were to increase the number of attributes inherited this would become more of a problem.